### PR TITLE
Add global variable to pass an index of all posts to the singlePost function

### PR DESCRIPTION
**Pull request by marcusguttenplan**
Add a `for` loop to `allPosts()` to push an index of posts to a global array, and change `singlePost()`'s `buffer` to pass the `index` variable as well as the original `post` as an object to display a list of all posts from the `single.hbs` template

### CODE REVIEW

**Comment by marcusguttenplan** [comment]
This is a great tool, thank you for your time and effort!

I just wanted to submit a PR because this was a feature I really wanted and hacked in, figured it was worth sharing just in case it may be found helpful by others. Feel free to reject it if it's out of scope or not up to snuff!

**Comment by JosephDavis** [comment]
Thanks for your contribution! This sounds like a good idea. I will take a look at the code changes this evening.

**Comment by marcusguttenplan** [comment]
Ran into an issue last night with the `{{#index}}` template tag being rendered again with every handlebars recompile, so if you see an issue with async and the changes or any bugs I'd love to refactor it into something production-quality.

File: index.js [hunk 1]
```diff
@@ -5,6 +5,8 @@ var File = require('vinyl');
 var path = require('path');
 var hljs = require('highlight.js');
 
+var pageIndex = [];
```
Comment (JosephDavis) [inline index.js:8]: This `pageIndex` variable is redundant to the existing `posts` variable [here](https://github.com/JosephDavis/gulp-flat-blog/blob/edaaa4caaa05269ef3f2ea7c29c467c3d4b61bae/index.js#L33).

File: index.js [hunk 2]
```diff
@@ -73,6 +75,12 @@ function gulpFlatBlog(templates) {
     var file = new File({
       path: 'index.html'
     });
+    for (i in posts) {
```
Comment (JosephDavis) [inline index.js:78]: You mentioned running into duplicated posts and added this check. The duplicates are a result of the location of the `pageIndex` variable, which is outside of the `gulpFlatBlog` function. In theory, if someone had 2 configurations for `gulpFlatBlog` in their gulpfile (for example, blogs in separate directories), this means posts from both would be accumulated in the `pageIndex` array. The existing `posts` variable is scoped within the `gulpFlatBlog` function, but still accessible from both `singlePost` and `allPosts`.

File: index.js [hunk 3]
```diff
@@ -53,7 +55,7 @@ function gulpFlatBlog(templates) {
         post.slug = file.relative.substring(0, file.relative.lastIndexOf('.'));
       }
 
-      file.contents = new Buffer(templates.single(post));
+      file.contents = new Buffer(templates.single({post: post, index: pageIndex}));
```
Comment (JosephDavis) [inline index.js:58]: I think some users may find using `index` as the name confusing in their templates, as several templating engines use `index` as some kind of keyword (e.g. `@index` in handlebars). May I suggest `post` and `posts`?

**Review by JosephDavis — CHANGES_REQUESTED** [review]
If you adjust the code according to the review comments I have added, I will gladly accept your pull request. Mainly, I just want to make sure the array of posts is scoped within the `gulpFlatBlog` function (to prevent accidentally sharing posts between separate configurations), and prevent any confusion in templates.

Off the top of my head, this might be easiest accomplished by accumulating the posts first in `singlePost`, then writing all the individual posts and the main "index" template page in `allPosts` (that way the posts can be added to the array and sorted before outputting any files).

**Comment by marcusguttenplan** [comment]
Thank you for the comments!

I originally tried to simply pass the `posts` object to `file.contents = new Buffer(templates.single({post: post, posts: posts}));` in `singlePost()`, but because of the callback it would only return the items that had already been pushed into `posts`.

Working from your comment about writing all individual posts as well as the index from `allPosts()`, I refactored along the lines of:
```
for (i in posts) {
        file.contents = new Buffer(templates.single({post: posts[i], posts: posts}));
        file.path = path.join(file.base, posts[i].slug + '.html');
        this.push(file);
    }
```

But this of course conflicts with the `file` variable, and doesn't have access to the file argument from `singlePost()`. I found a hacky solution by defining a second separate `file2` variable and got it to successfully pass the `posts` to the single template, but it only seems to work at the first or last entries of the `posts` array.

Let me know if you have any advice, or again feel free to close if this is too much.

[Here's a gist with some sketches.](https://gist.github.com/marcusguttenplan/98bf83832d9c3e4663dba230b06a1eef)

**Comment by marcusguttenplan** [comment]
I'm going to go ahead and close this PR. I found a solution that works for me, with multiple instances of gulp-flat-blog, but it is super hacky. I couldn't figure out a good way to get it all into a separate callback. Ended up pushing each single rendered file from the allPosts function, and used singlePost to just create the frontmatter and index. Not ideal, but works for my needs!

For reference:
```
function gulpFlatBlog(templates) {
  if (!templates.single || !templates.index) {
    throw new PluginError('gulp-flat-blog', 'Missing template.');
  }

  var posts = [];

  function singlePost(file, encoding, callback) {

    if (file.isStream()) {
      this.emit('error', new PluginError('gulp-flat-blog',
        'Streams are not supported.'));
      return callback();
    }

    if (file.isBuffer()) {


      var post = frontMatter(file.contents.toString());
      post.body = markdownit.render(post.body);

      for (var key in post.attributes) {
          post[key] = post.attributes[key];
      }
      delete post.attributes;

      if (!post.hasOwnProperty('slug')) {
        post.slug = file.relative.substring(0, file.relative.lastIndexOf('.'));
      }

      post.createdAt = file.stat.birthtime;
      post.updatedAt = file.stat.mtime;
      posts.push(post);

      file.contents = new Buffer(templates.single({post: post}));
      file.path = path.join(file.base, post.slug + '.html');

    }
    // this.push(file);
    callback();
  }


  function allPosts(callback) {

    posts.sort(function(lhs, rhs) {
      return new Date(rhs.date ? rhs.date : rhs.pubdate) - new Date(
        lhs.date ? lhs.date : lhs.pubdate);
    });

    var file = new File({
      path: 'index.html'
    });


    file.contents = new Buffer(templates.index({
      posts: posts
    }));

    var vars = {};

    for (i in posts) {
        vars['file' + i] = new File({path: 'single.html'});

        vars['file' + i].contents = new Buffer(templates.single({post: posts[i], posts: posts}));
        vars['file' + i].path = path.join(vars['file' + i].base, posts[i].slug + '.html');
        this.push(vars['file' + i]);
    }

    this.push(file);
    callback();
  }

  return through.obj(singlePost, allPosts);
};
```

**Comment by JosephDavis** [comment]
Thank you again. Unfortunately, with my other commitments and work, I haven't had time to set aside for open source projects. I will take a look at some point.

**Pull request closed without merging**